samples/solutions/stingy_connection.py

```diff
@@ -0,0 +1,275 @@
+            exceptions.ConnectionDroppedError,
+            exceptions.NoConnectionError,
+            exceptions.ClientError,  # TODO Only TLSEsxhangeError is actually worth retrying
+        ]
+
+        self.message_queue = asyncio.Queue()
+        self.message_list = {}
+        self.task_list = []
+        self.device_client = None
+
+    async def create_client(self, conn_str):
+        try:
+            # Create a Device Client
+            self.device_client = IoTHubDeviceClient.create_from_connection_string(conn_str)
+            # Attach the connection state handler
+            self.device_client.on_connection_state_change = self.handle_on_connection_state_change
+        except Exception as e:
+            self.log_error_and_print(
+                "Caught exception while trying to attach handler : {}".format(get_type_name(e))
+            )
+
+    async def handle_on_connection_state_change(self):
+        self.log_info_and_print(
+            "handle_on_connection_state_change fired. Connected status : {}".format(
+                self.device_client.connected
+            )
+        )
+        if self.device_client.connected:
+            self.log_info_and_print("Connected connected_event is set...")
+        else:
+            self.log_info_and_print("Disconnected connected_event is set...")
+
+    def is_retryable(self, exc):
+        if type(exc) in self.transient_errors:
+            return True
+        else:
+            return False
+
+    def log_error_and_print(self, s):
+        logger.error(s)
+        print(s)
+
+    def log_info_and_print(self, s):
+        logger.info(s)
+        print(s)
+
+    async def enqueue_message(self):
+        message_id = 0
+        while True:
+            msg = Message("current wind speed ")
+            msg.message_id = message_id
+            msg.content_type = "application/json"
+            self.log_info_and_print("Created a message...")
+            self.message_queue.put_nowait(msg)
+            self.message_list[message_id] = msg
+            # time between 10 seconds and less than 30 minutes.
+            randint = random.randint(UPPER_LIMIT_OF_ENQUEUEING, UPPER_LIMIT_OF_ENQUEUEING)
+            self.log_info_and_print(
+                "Will sleep for {} seconds and then enqueue messages".format(randint)
+            )
+            await asyncio.sleep(randint)
+            message_id += 1
+
+    async def do_all_tasks_and_disconnect(self):
+        self.log_info_and_print("Current qsize is: {}".format(self.message_queue.qsize()))
+        while not self.message_queue.empty():
+            msg = await self.message_queue.get()
+            task = asyncio.create_task(self.device_client.send_message(msg))
+            self.task_list.append(task)
+            self.message_queue.task_done()
+        try:
+            if self.device_client.connected:
+                self.log_info_and_print("sending messages...")
+                await asyncio.gather(*self.task_list)
```

Comment:
>    await asyncio.gather(*self.task_list) [](http://example.com/codeflow?start=12&length=41)

What happens if one of these tasks raises an exception? The tasks all stay in `self.task_list`, but you don't try to re-send anything that failed.  I also think trying to gather the same task twice will raise an exception, so keeping the completed tasks for the next call to gather will cause gather to fail every time (after the first failure)

Reply:
not resending just clearing task list in case sending message runs into an exception